File: benchmarks/bench.cpp
// C++ Heavy Benchmark - CPU, Memory, Concurrency
#include <iostream>
#include <chrono>
#include <thread>
#include <vector>
#include <cmath>

bool is_prime(long n) {
    if (n < 2) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;
    for (long i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}

long count_primes(int limit) {
    long count = 0;
    for (int i = 2; i < limit; i++) {
        if (is_prime(i)) count++;
    }
    return count;
}

long memory_stress(int allocs) {
    long total = 0;
    for (int i = 0; i < allocs; i++) {
        auto* ptr = new char[1024];
        total += i;
        delete[] ptr;
    }
    return total;
}

long fib(int n) {
    if (n <= 1) return n;
    long a = 0, b = 1;
    for (int i = 2; i <= n; i++) {
        long c = a + b;
        a = b;
        b = c;
    }
    return b;
}

void worker() {
    count_primes(10000);
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    
    // CPU Test
    long primes = count_primes(100000);
    
    // Memory Test
    long mem = memory_stress(10000);
    
    // Compute Test
    long f = fib(40);
    
    // Concurrency Test: 4 threads
    std::vector<std::thread> threads;
    for (int i = 0; i < 4; i++) {
        threads.emplace_back(worker);
    }
    for (auto& t : threads) {
        t.join();
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
    
    std::cout << "C++ Results:" << std::endl;
    std::cout << "  Primes: " << primes << std::endl;
    std::cout << "  Fib(40): " << f << std::endl;
    std::cout << "  Time: " << elapsed << " ms" << std::endl;
    
    return 0;
}

File: benchmarks/bench.rs
// Rust Heavy Benchmark - CPU, Memory, Concurrency
use std::time::Instant;
use std::thread;

fn is_prime(n: i64) -> bool {
    if n < 2 { return false; }
    if n == 2 { return true; }
    if n % 2 == 0 { return false; }
    let mut i = 3;
    while i * i <= n {
        if n % i == 0 { return false; }
        i += 2;
    }
    true
}

fn count_primes(limit: i64) -> i64 {
    (2..limit).filter(|&n| is_prime(n)).count() as i64
}

fn memory_stress(allocs: i32) -> i64 {
    let mut total: i64 = 0;
    for i in 0..allocs {
        let v: Vec<u8> = vec![0u8; 1024];
        total += i as i64;
        drop(v);
    }
    total
}

fn fib(n: i64) -> i64 {
    if n <= 1 { return n; }
    let (mut a, mut b) = (0i64, 1i64);
    for _ in 2..=n {
        let c = a + b;
        a = b;
        b = c;
    }
    b
}

fn main() {
    let start = Instant::now();
    
    // CPU Test
    let primes = count_primes(100000);
    
    // Memory Test
    let mem = memory_stress(10000);
    
    // Compute Test
    let f = fib(40);
    
    // Concurrency Test: 4 threads
    let handles: Vec<_> = (0..4).map(|_| {
        thread::spawn(|| count_primes(10000))
    }).collect();
    
    for h in handles {
        let _ = h.join();
    }
    
    let elapsed = start.elapsed().as_millis();
    
    println!("Rust Results:");
    println!("  Primes: {}", primes);
    println!("  Fib(40): {}", f);
    println!("  Time: {} ms", elapsed);
}
